src/main/java/com/ryanharter/auto/value/moshi/AutoValueMoshiAdapterFactoryProcessor.java

```diff
@@ -128,13 +130,26 @@ private TypeSpec createJsonAdapterFactory(
 
     for (int i = 0; i < elements.size(); i++) {
       Element element = elements.get(i);
+      TypeName elementTypeName = TypeName.get(element.asType());
+      CodeBlock typeBlock = CodeBlock.of("$N", type);
+
+      if(elementTypeName instanceof ParameterizedTypeName) {
+        elementTypeName = ((ParameterizedTypeName) elementTypeName).rawType;
+        typeBlock = CodeBlock.of("$N instanceof $T && (($T) $N).getRawType()", type, ParameterizedType.class, ParameterizedType.class, type);
```

Comment:
If more then one generic type is declared, wont that create multiple `if (type instanceof ParameterizedType /* && so on */)` blocks? seams kinda redundant to me.
 May be it makes sense to try to optimize the statement, and go for something like this:

```
@Override public JsonAdapter<?> create(Type type, Set<? extends Annotation> annotations, Moshi moshi) {
  if (type instanceof ParameterizedType) {
    Type rawType = (($T) $N).getRawType();
    if (rawType.equals(Foo.class)) {
      // return adapter
    } else if (rawType.equals(Bar.class)) {
      // return adapter
    }
    return null;
  }

  if (type.equals(OtherValueClass.class) {
  } else if (/* and so on */) {
  }

  return null;
}
```